api/mfx_dispatch/windows/include/mfx_dispatcher.h

```diff
@@ -218,4 +218,76 @@ FUNCTION_DESCRIPTION APIFunc[eVideoFuncTotal];
 
 extern const
 FUNCTION_DESCRIPTION APIAudioFunc[eAudioFuncTotal];
+
+
+#if (defined(_WIN64) || defined(_WIN32)) && (MFX_VERSION >= MFX_VERSION_NEXT)
+enum mfxComponentType
+{
+    ENCODE = 1,
+    DECODE = 2,
+    VPP    = 3
+};
+
+struct mfxInfoCodec
+{
+    mfxU32       reserved[7];
+    mfxU16       LowPower;
+    mfxU16       reserved1;
+
+    mfxFrameInfo FrameInfo;
+
+    mfxU32       CodecId;
+    mfxU16       CodecProfile;
+    mfxU16       CodecLevel;
+    mfxU16       reserved2;
+    mfxU16       TargetUsage;
+
+    mfxU16       reserved3[12];
+};
+
+struct mfxAdapterRequirements
+{
+    union
+    {
+        mfxInfoCodec Codec;
+        mfxInfoVPP   VPP;
+    };
+
+    mfxExtBuffer** ExtParam;
+    mfxU16         NumExtParam;
+
+    mfxU16         reserved[7];
+};
+
+struct mfxComponentInfo
+{
+    mfxComponentType       Type;
+    mfxAdapterRequirements Requirements;
+
+    mfxU16                 reserved[4];
+};
+
+struct mfxAdapterInfo
+{
+    mfxPlatform Platform;
+    mfxU32      Number;
+
+    mfxU16      reserved[14];
+};
+
+struct mfxAdaptersInfo
+{
+    mfxAdapterInfo * Adapters;
+    mfxU32           NumAlloc;
+    mfxU32           NumActual;
+
+    mfxU16           reserved[4];
+};
+
+mfxStatus MFXQueryAdapters(mfxComponentInfo* input_info, mfxAdaptersInfo* adapters);
+mfxStatus MFXQueryAdaptersDecode(mfxBitstream* bitstream, mfxU32 codec_id, mfxAdaptersInfo* adapters);
+mfxStatus MFXQueryAdaptersNumber(mfxU32* num_adapters);
```

Comment:
This API follows approach: I have something, do you support that? That's a question not everyone bothers to actually ever ask, hence lower number of adoptions than there could be:). Alternative approach which I believe suites better is: give me what you support and I will check whether I can use it?

From the perspective of the 2nd API approach there would be 1 or 2 primary functions:
```
mfxStatus MFXQueryAdaptersNumber(mfxU32* num_adapters);
mfxStatus MFXQueryAdaptersInfo(mfxU32 num_adapters, mfxAdapterInfo* adapters);
```

After that we can have any number of helper functions which utilize the above query API and search thru it.

Mind that with this approach you immediately get opportunity to write very simple sample application which queries all adapters and reports back their caps.

Reply:
I'm absolutely agree with you. But we have to define capabilities somehow. We need to define some general enough struct that can serve as description for supported features. It might happen that such struct would be giant and thus not easy to manipulate. And the second and most hard thing - we need to define those caps. Actually current API is extendable to support such approach also. We can add pointer to capabilities to ```mfxAdapterInfo```. So we can use it as you suggested in following way via proposed API: pass ```nullptr``` instead of requirements and attach capabilities struct to each ```mfxAdapterInfo```. Then MSDK lib will fill those structs with information about supported features for every adapter. I think we may introduce such feature later after merge of this API, because it will take some time to design proper caps struct and implementing such reporting mechanism.